README.org
#+TITLE: Deep Learning Training with Multi-Party Computation

Small set of scripts for training MNIST and CIFAR10 with a number of
networks using [[https://github.com/data61/MP-SPDZ/][MP-SPDZ]]. This version underlies the figures in
<https://arxiv.org/abs/2107.00501>. For the version underlying
<https://eprint.iacr.org/2020/1330> and
<https://arxiv.org/abs/2011.11202>, see the =v1= branch.

** Installation (with Docker)


Run the following  from this directory to build a Docker container:

: $ docker build .

=Dockerfile= contains some commented out tests at the end.

** Running locally

After setting everything up, you can use this script to run the
computation:

: $ ./run-local.sh <protocol> <net> <round> <n_threads> <n_epochs> <precision> [<further>]

The options are as follows:

- =protocol= is one of =emul= (emulation), =sh2= (semi-honest
  two-party computation) =sh3= (semi-honest three-party computation,
  =mal3= (malicious three-party computation), =mal4= (malicious
  four-party computation), =dm3= (dishonest-majority semi-honest
  three-party computation), =sh10= (semi-honest ten-party
  computation), =dm10= (dishonest-majority semi-honest ten-party
  computation).  All protocols assume an honest majority unless stated
  otherwise.
- =net= is the network (A-D for MNIST, alex for Falcon AlexNet on
  CIFAR10, and new_alex for a more sophisticated AlexNet-like network on
  CIFAR10).
- =round= is the kind of rounding, =prob= for probabilistic and =near= for
  nearest.
- =n_threads= is the number of threads per party.
- =n_epochs= is the number of epochs.
- =precision= is the precision (in bits) after the decimal point.

The following options can be given in addition.

- =rate<rate>= the learning rate, e.g., =rate.1= for a learning rate of 0.1.
- =adam=, =adamapprox=, or =amsgrad= to use Adam, Adam with less
  precise approximation of inverse square root, or AMSgrad instead of
  SGD.

For example,

: $ ./run-local.sh emul D prob 2 20 32 adamapprox

runs 20 epochs of training network D in the emulation with two threads,
probabilistic rounding, 32-bit precision, and Adam with less precise
inverse square root approximation.

** Running remotely

You need to set up hosts that run SSH and have all higher TCP ports
open between each other. We have used =c5.9xlarge= instances in the
same AWS zone and hence 36 threads. The hosts have to run Linux with a
glib not older than Ubuntu 18.04 (2.27), which is the case for Amazon
Linux 2. Honest-majority protocols require three hosts while
dishonest-majority protocols require two.

With Docker, you can run the following script to set up host names,
user name and SSH RSA key. We do *NOT* recommend running it outside
Docker because it might overwrite an existing RSA key file.

: $ ./setup-remote.sh

Without Docker, familiarise yourself with SSH configuration options
and SSH keys. You can use =ssh_config= and the above script to find
out the requirements. =HOSTS= has to contain the hostnames separated
by whitespace.

After setting up, you can the following using the same options as
above:

: $ ./run-remote.sh <protocol> <net> <round> <n_threads> <n_epochs> <precision> [<further>]

For example,

: $ ./run-remote.sh sh3 A near 1 1 16 rate.1

runs one epoch of training network A with semi-honest three-party
computation, one thread, and nearest rounding, 16-bit precision,
and SGD with rate 0.1.

** Cleartext training

=train.py= allows to run the equivalent cleartext training with Tensorflow
(using floating-point instead of fixed-point representation).
Simply run the following:

: $ ./train.py <network> <n_epochs> [<optimizer>] [<further>]

=<optimizer>= is either a learning rate for SGD, or =adam= or =amsgrad=
followed by a learning rate. =<further>= is any combination of =dropout=
to add a Dropout layer (only with network C) and =fashion= for
Fashion MNIST.

** Data format and scripts

=full.py= processes the whole MNIST dataset, and =full01.py= processes
two digits to allow logistic regression and similar. By
default the digits are 0 and 1, but you can request any pair. The
following processes the digits 4 and 9, which are harder to
distinguish:

: $ ./full01.py 49

Both output to stdout in the format that =mnist_*.mpc= expect in the
input file for party 0, that is: training set labels, training set
data (example by example), test set labels, test set data (example by
example), all as whitespace-seperated text. Labels are stored as
one-hot vectors for the whole dataset and 0/1 for logistic regression.

The scripts expect the unzipped MNIST dataset in the current
directory. Run =download.sh= to download the MNIST dataset.


MP-SPDZ/CHANGELOG.md
The changelog explains changes pulled through from the private development repository. Bug fixes and small enhancements are committed between releases and not documented here.

## 0.3.7 (August 14, 2023)

- Path Oblivious Heap (@tskovlund)
- Adjust batch and bucket size to program
- Direct communication available in more protocols
- Option for seed in fake preprocessing (@strieflin)
- Lower memory usage due to improved register allocation
- New instructions to speed up CISC compilation
- Protocol implementation example
- Fixed security bug: missing MAC checks in multi-threaded programs
- Fixed security bug: race condition in MAC check
- Fixed security bug: missing shuffling check in PS mod 2^k and Brain
- Fixed security bug: insufficient drowning in pairwise protocols

## 0.3.6 (May 9, 2023)

- More extensive benchmarking outputs
- Replace MPIR by GMP
- Secure reading of edaBits from files
- Semi-honest client communication
- Back-propagation for average pooling
- Parallelized convolution
- Probabilistic truncation as in ABY3
- More balanced communication in Shamir secret sharing
- Avoid unnecessary communication in Dealer protocol
- Linear solver using Cholesky decomposition
- Accept .py files for compilation
- Fixed security bug: proper accounting for random elements

## 0.3.5 (Feb 16, 2023)

- Easier-to-use machine learning interface
- Integrated compilation-execution facility
- Import/export sequential models and parameters from/to PyTorch
- Binary-format input files
- Less aggressive round optimization for faster compilation by default
- Multithreading with client interface
- Functionality to protect order of specific memory accesses
- Oblivious transfer works again on older (pre-2011) x86 CPUs
- clang is used by default

## 0.3.4 (Nov 9, 2022)

- Decision tree learning
- Optimized oblivious shuffle in Rep3
- Optimized daBit generation in Rep3 and semi-honest HE-based 2PC
- Optimized element-vector AND in SemiBin
- Optimized input protocol in Shamir-based protocols
- Square-root ORAM (@Quitlox)
- Improved ORAM in binary circuits
- UTF-8 outputs

## 0.3.3 (Aug 25, 2022)

- Use SoftSpokenOT to avoid unclear security of KOS OT extension candidate
- Fix security bug in MAC check when using multithreading
- Fix security bug to prevent selective failure attack by checking earlier
- Fix security bug in Mama: insufficient sacrifice.
- Inverse permutation (@Quitlox)
- Easier direct compilation (@eriktaubeneck)
- Generally allow element-vector operations
- Increase maximum register size to 2^54
- Client example in Python
- Uniform base OTs across platforms
- Multithreaded base OT computation
- Faster random bit generation in two-player Semi(2k)

## 0.3.2 (May 27, 2022)

- Secure shuffling
- O(n log n) radix sorting
- Documented BGV encryption interface
- Optimized matrix multiplication in dealer protocol
- Fixed security bug in homomorphic encryption parameter generation
- Fixed security bug in Temi matrix multiplication

## 0.3.1 (Apr 19, 2022)

- Protocol in dealer model
- Command-line option for security parameter
- Fixed security bug in SPDZ2k (see Section 3.4 of [the updated paper](https://eprint.iacr.org/2018/482))
- Ability to run high-level (Python) code from C++
- More memory capacity due to 64-bit addressing
- Homomorphic encryption for more fields of characteristic two
- Docker container

## 0.3.0 (Feb 17, 2022)

- Semi-honest computation based on threshold semi-homomorphic encryption
- Batch normalization backward propagation
- AlexNet for CIFAR-10
- Specific private output protocols
- Semi-honest additive secret sharing without communication
- Sending of personal values
- Allow overwriting of persistence files
- Protocol signature in persistence files

## 0.2.9 (Jan 11, 2022)

- Disassembler
- Run-time parameter for probabilistic truncation error
- Probabilistic truncation for some protocols computing modulo a prime
- Simplified C++ interface
- Comparison as in [ACCO](https://dl.acm.org/doi/10.1145/3474123.3486757)
- More general scalar-vector multiplication
- Complete memory support for clear bits
- Extended clear bit functionality with Yao's garbled circuits
- Allow preprocessing information to be supplied via named pipes
- In-place operations for containers

## 0.2.8 (Nov 4, 2021)

- Tested on Apple laptop with ARM chip
- Restore trusted client interface
- Directly accessible softmax function
- Signature in preprocessing files to reduce confusing errors
- Improved error messages for connection issues
- Documentation of low-level share types and protocol pairs

## 0.2.7 (Sep 17, 2021)

- Optimized matrix multiplication in Hemi
- Improved client communication
- Private integer division as per [Veugen and Abspoel](https://doi.org/10.2478/popets-2021-0073)
- Compiler option to translate some Python control flow instructions
  to run-time instructions
- Functionality to break out of run-time loops
- Run-time range check of data structure accesses
- Improved documentation of network infrastructure

## 0.2.6 (Aug 6, 2021)

- [ATLAS](https://eprint.iacr.org/2021/833)
- Keras-like interface
- Iterative linear solution approximation
- Binary output
- HighGear/LowGear key generation for wider range of parameters by default
- Dabit generation for smaller primes and malicious security
- More consistent type model
- Improved local computation
- Optimized GF(2^8) for CCD
- NTL only needed for computation with GF(2^40)
- Virtual machines suggest compile-time optimizations
- Improved documentation of types

## 0.2.5 (Jul 2, 2021)

- Training of convolutional neural networks
- Bit decomposition using edaBits
- Ability to force MAC checks from high-level code
- Ability to close client connection from high-level code
- Binary operators for comparison results
- Faster compilation for emulation
- More documentation
- Fixed bug in dense layer back-propagation
- Fixed security bug: insufficient LowGear secret key randomness
- Fixed security bug: skewed random bit generation

## 0.2.4 (Apr 19, 2021)


train.py
#!/usr/bin/python3

# -*- coding: utf-8 -*-
"""tensorflow/datasets

Automatically generated by Colaboratory.

Original file is located at
    https://colab.research.google.com/github/tensorflow/datasets/blob/master/docs/keras_example.ipynb

# Training a neural network on MNIST with Keras

This simple example demonstrate how to plug TFDS into a Keras model.

Copyright 2020 The TensorFlow Datasets Authors, Licensed under the Apache License, Version 2.0

<table class="tfo-notebook-buttons" align="left">
  <td>
    <a target="_blank" href="https://www.tensorflow.org/datasets/keras_example"><img src="https://www.tensorflow.org/images/tf_logo_32px.png" />View on TensorFlow.org</a>
  </td>
  <td>
    <a target="_blank" href="https://colab.research.google.com/github/tensorflow/datasets/blob/master/docs/keras_example.ipynb"><img src="https://www.tensorflow.org/images/colab_logo_32px.png" />Run in Google Colab</a>
  </td>
  <td>
    <a target="_blank" href="https://github.com/tensorflow/datasets/blob/master/docs/keras_example.ipynb"><img src="https://www.tensorflow.org/images/GitHub-Mark-32px.png" />View source on GitHub</a>
  </td>
</table>
"""

import tensorflow.compat.v2 as tf
import tensorflow_datasets as tfds
import sys
import re

tfds.disable_progress_bar()
tf.enable_v2_behavior()

"""## Step 1: Create your input pipeline

Build efficient input pipeline using advices from:
* [TFDS performance guide](https://www.tensorflow.org/datasets/performances)
* [tf.data performance guide](https://www.tensorflow.org/guide/data_performance#optimize_performance)

### Load MNIST

Load with the following arguments:

* `shuffle_files`: The MNIST data is only stored in a single file, but for larger datasets with multiple files on disk, it's good practice to shuffle them when training.
* `as_supervised`: Returns tuple `(img, label)` instead of dict `{'image': img, 'label': label}`
"""

dataset = 'fashion_mnist' if 'fashion' in sys.argv else 'mnist'

(ds_train, ds_test), ds_info = tfds.load(
    dataset,
    split=['train', 'test'],
    shuffle_files=True,
    as_supervised=True,
    with_info=True,
)

"""### Build training pipeline

Apply the following transormations:

* `ds.map`: TFDS provide the images as tf.uint8, while the model expect tf.float32, so normalize images
* `ds.cache` As the dataset fit in memory, cache before shuffling for better performance.<br/>
__Note:__ Random transformations should be applied after caching
* `ds.shuffle`: For true randomness, set the shuffle buffer to the full dataset size.<br/>
__Note:__ For bigger datasets which do not fit in memory, a standard value is 1000 if your system allows it.
* `ds.batch`: Batch after shuffling to get unique batches at each epoch.
* `ds.prefetch`: Good practice to end the pipeline by prefetching [for performances](https://www.tensorflow.org/guide/data_performance#prefetching).
"""

def normalize_img(image, label):
  """Normalizes images: `uint8` -> `float32`."""
  return tf.cast(image, tf.float32) / 255., label

ds_train = ds_train.map(
    normalize_img, num_parallel_calls=tf.data.experimental.AUTOTUNE)
ds_train = ds_train.cache()
ds_train = ds_train.shuffle(ds_info.splits['train'].num_examples)
ds_train = ds_train.batch(20)
ds_train = ds_train.prefetch(tf.data.experimental.AUTOTUNE)

"""### Build evaluation pipeline

Testing pipeline is similar to the training pipeline, with small differences:

 * No `ds.shuffle()` call
 * Caching is done after batching (as batches can be the same between epoch)
"""

ds_test = ds_test.map(
    normalize_img, num_parallel_calls=tf.data.experimental.AUTOTUNE)
ds_test = ds_test.batch(10)
ds_test = ds_test.cache()
ds_test = ds_test.prefetch(tf.data.experimental.AUTOTUNE)

"""## Step 2: Create and train the model

Plug the input pipeline into Keras.
"""

print(ds_train)
print(ds_test)


network = 'A'
n_epochs = 1
lr = 0.01
adam = False
amsgrad = False

if len(sys.argv) > 1:
  network = sys.argv[1]
if len(sys.argv) > 2:
  n_epochs = int(sys.argv[2])
if len(sys.argv) > 3:
  adam = re.match('adam(.*)', sys.argv[3])
  amsgrad = re.match('amsgrad(.*)', sys.argv[3])
  if adam or amsgrad:
    try:
      lr = float((adam or amsgrad).group(1))
    except:
      lr = .001
    if adam:
      print('use Adam with lr', lr)
    else:
      print('use AMSGrad with lr', lr)
  else:
    lr = float(sys.argv[3])

if network == 'A':
  layers = [
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(128,activation='relu'),
    tf.keras.layers.Dense(128,activation='relu'),
    tf.keras.layers.Dense(10, activation='softmax')
  ]
elif network == 'D':
  layers = [
    tf.keras.layers.Conv2D(5, 5, 2, 'same', activation='relu'),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(100, activation='relu'),
    tf.keras.layers.Dense(10, activation='softmax')
  ]
  model = tf.keras.models.Sequential(layers)
elif network == 'B':
  layers = [
    tf.keras.layers.Conv2D(16, 5, 1, 'valid', activation='relu'),
    tf.keras.layers.MaxPooling2D(2),
    tf.keras.layers.Conv2D(16, 5, 1, 'valid', activation='relu'),
    tf.keras.layers.MaxPooling2D(2),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(100, activation='relu'),
    tf.keras.layers.Dense(10, activation='softmax')
  ]
elif network == 'C':
  layers = [
    tf.keras.layers.Conv2D(20, 5, 1, 'valid', activation='relu'),
    tf.keras.layers.MaxPooling2D(2),
    tf.keras.layers.Conv2D(50, 5, 1, 'valid', activation='relu'),
    tf.keras.layers.MaxPooling2D(2),
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(500, activation='relu'),
    tf.keras.layers.Dense(10, activation='softmax')
  ]
  if 'dropout' in sys.argv:
    layers.insert(-2, tf.keras.layers.Dropout(0.5))
elif network == 'small_mnist':
  layers = [
    tf.keras.layers.Flatten(),
    tf.keras.layers.Dense(64, activation='relu'),
    tf.keras.layers.Dense(10),
  ]
else:
  raise Exception('unknown network: ' + network)

print (layers)
model = tf.keras.models.Sequential(layers)

if adam or amsgrad:
  optim = tf.keras.optimizers.Adam(lr, amsgrad=amsgrad)
else:
  optim = tf.keras.optimizers.SGD(momentum=0.9, learning_rate=lr,)

model.compile(
  # loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
  loss=tf.keras.losses.MeanSquaredError(reduction="auto", name="mean_squared_error"),
  metrics=[tf.keras.metrics.SparseCategoricalAccuracy()],
  optimizer=optim
)

for i in range(1):
  history = model.fit(
    ds_train,
    epochs=n_epochs,
    validation_data=ds_test,
  )

  out = open('log-' + '-'.join(sys.argv[1:]), 'w')
  for i in range(n_epochs):
    out.write(str(i) + ' ')
    for x in history.history:
      out.write('%s: %.4f ' % (x, history.history[x][i]))
    out.write('\n')
